### ISSUE TITLE
Warn about find without a path parameter.

### ISSUE THREAD

**Issue by Herst** [koalaman/shellcheck#627]
I thought the path parameter in find was optional and if left out the current path (".") was be implied but apparently that is not everywhere the case and not according to the POSIX standard (and apparently also not e.g. in a version of ```bash``` for OS X from what I heard from someone using my script).
So it would be great if `shellcheck` would warn about:
```sh
#!/bin/sh
find -type f
```
or
```bash
#!/bin/bash
find -type f
```

**Comment by kurahaupo** [comment]
It doesn't matter which shell is specified, unless it's one that has `find`
as a built-in (and there are none that do AFAIK, so it's moot).

**Comment by kurahaupo** [comment]
While we're at it, we should probably warn when `mv`, `cp` or `ln` have
fewer than 2 filename args (when they're statically countable).

Personally I'd like a UUoC warning too -- or just include `cat` in the list
of commands that must have at least 2 filename args -- but that's probably
going a bit far for most people.

**Comment by koalaman** [comment]
ShellCheck currently doesn't have a notion of which OS the script will run on, and as such it generally doesn't check portability of external utilities.

There should be a os=posix directive or similar for types of warnings.

**Comment by bittorf** [comment]
related e.g. #567 and #531

**Comment by Herst** [comment]
> ShellCheck currently doesn't have a notion of which OS the script will run on, and as such it generally doesn't check portability of external utilities.
OK, I did not know that, thanks!
I want to note that I am always quite sure what is what is not considered built-in and what not, also because I have a lot to do with `busybox` where a lot can basically be built-in into the executable but is still not necessarily considered built-in into the "shell" [according to commands like `type`, also depends on what compile settings you are using it really searches the path for commands or prefers to call commands inside it right away] and e.g. "[" isn't part of the ash shell there but is always in the executable.
Also since `shellcheck` also advocates the use of such external utilities (e.g. `find` output is favored over the one from `ls` for handling bay scripts) it would make sense to add support to this for shellcheck at some point even if it would mean a lot of work.

**Comment by brother** [comment]
If you are not sure but want to check if it's builtin or not you could use the which - if it's available ofc. The catch 22 situation is very nice here =)

```
brother ~$ which which
/bin/which
brother ~$ ls -l /bin/which
-rwxr-xr-x 1 root root 946 jan 26 19:17 /bin/which
brother ~$ which read
brother ~$ which find
/usr/bin/find
...
root@172.21.13.70:~# which touch
/bin/touch
root@172.21.13.70:~# ls -l /bin/touch
lrwxrwxrwx 1 root root 12 Feb 27 22:58 /bin/touch -> /bin/busybox
```
and so on.

**Comment by kurahaupo** [comment]
No no no noooo! Do *NOT* recommend `which`; it's a bad solution looking for
a question that already has a better answer.

The `which` command will report where *it* can find commands, which means
you will have unreliable results for (or possibly always completely missing)
shell keywords, shell built-ins, shell functions, dynamic loaders, and shell
aliases. Even if your version of `which` can read `~/.bashrc` it may still
give wrong answers, because defining aliases and functions may be dependent
on context that it can't compute, or use a syntax it doesn't understand.

The `type` command is built into the shell and will tell you what the shell
will actually use, rather than what some external utility will second-guess
about the shell.

The only reliable version of `which` is to define it a shell function:
`which() { type "$@"; }` but if you're going to do that, you might as well
just type `type` to begin with.

**Comment by brother** [comment]
Just to make it clear: I did not propose to use which for real. It was a poke towards the main problem that you do not know the environment. Using which when you do not know if which is available is a broken problem.

**Comment by koalaman** [comment]
b0dae06 adds an info message for `find` on all systems, without trying to solve the grander problem of differences in userspace.
